python/cuml/cuml/experimental/accel/module_accelerator.py

```diff
@@ -0,0 +1,664 @@
+            # another thread might modify it.
+            # Modification has to happen with the lock held for the
+            # duration, so if someone else has modified things, then
+            # we block trying to acquire the lock (hence it is safe to
+            # release the lock after reading this value)
+            use_real = not loader._use_fast_lib
+        if not use_real:
+            # Only need to check the denylist if we're not turned off.
+            frame = sys._getframe()
+            # We cannot possibly be at the top level.
+            assert frame.f_back
+            calling_module = pathlib.PurePath(frame.f_back.f_code.co_filename)
+            use_real = _caller_in_denylist(
+                calling_module, tuple(loader._denylist)
+            )
+        try:
+            if use_real:
+                return real[name]
+            else:
+                return wrapped_objs[real[name]]
+        except KeyError:
+            raise AttributeError(f"No attribute '{name}'")
+        except TypeError:
+            # real[name] is an unhashable type
+            return real[name]
+
+    @classmethod
+    def install(
+        cls,
+        destination_module: str,
+        fast_lib: str,
+        slow_lib: str,
+    ) -> Self | None:
+        # This grabs the global _import_ lock to avoid concurrent
+        # threads modifying sys.modules.
+        # We also make sure that we finish installing ourselves in
+        # sys.meta_path before releasing the lock so that there isn't
+        # a race between our modification of sys.modules and someone
+        # else importing the slow_lib before we have added ourselves
+        # to the meta_path
+        with ImportLock():
+            logger.debug("Module Accelerator Install")
+            logger.debug(f"destination_module: {destination_module}")
+            logger.debug(f"fast_lib: {fast_lib}")
+            logger.debug(f"slow_lib: {slow_lib}")
+            logger.info("Non Estimator Function Dispatching disabled...")
+            if destination_module != slow_lib:
+                raise RuntimeError(
+                    f"Destination module '{destination_module}' must match"
+                    f"'{slow_lib}' for this to work."
+                )
+            mode = deduce_cuml_accel_mode(slow_lib, fast_lib)
+            if mode.use_fast_lib:
+                importlib.import_module(
+                    f".._wrappers.{mode.slow_lib}", __name__
+                )
+            try:
+                (self,) = (
+                    p
+                    for p in sys.meta_path
+                    if isinstance(p, cls)
+                    and p.slow_lib == mode.slow_lib
+                    and p.fast_lib == mode.fast_lib
+                )
+            except ValueError:
+                self = cls(mode.fast_lib, mode.slow_lib)
+                sys.meta_path.insert(0, self)
+            return self
+
+
+def disable_module_accelerator() -> contextlib.ExitStack:
```

Comment:
As a follow-on, we should be able to make this much faster using our global settings object. Not critical for the initial merge though.